docs/wiki/NOS3_cFS_Development.md

```diff
@@ -0,0 +1,114 @@
+    DataBuffer = (uint8_t *)malloc((1024) * sizeof(uint8_t));
+    /* Read the GPS data from the UART */
+    NAV_ReadAvailableData(DataBuffer, &DataLen);
+    GPSSerialiation GPSData = NAV_ParseOEM615Bestxyza(DataBuffer, DataLen);
+```
+
+The function `NAV_ReadAvailableData` is a wrapper for the low level UART calls to the OBC driver. The function can be seen in the following code excerpt. This code must include the hardware library, as seen in the first line of the excerpt, which includes the OBC drivers itself. The function calls beginning with `uart_` are from the OBC driver.
+
+```c
+#include ”hwlib.h”
+/* some code removed for readability see components/novatel_oem615/fsw/src/nav_app.c */
+static void NAV_ReadAvailableData(uint8 DataBuffer[], int32 *DataLen)
+{
+    int32 i = 0;
+    /* check how many bytes are waiting on the uart */
+    *DataLen = uart_bytes_available(NAV_UART.handle);
+    //OS_printf("NAV_ReadAvailableData(): gps messages waiting: %ld bytes\n", (long int)*DataLen);
+    /* declare an out buffer to hold that data */
+    if (*DataLen > 0)
+    {
+        /* grab the bytes */
+        uart_read_port(NAV_UART.handle, DataBuffer, *DataLen);
+    }
+    else
+    {
+        /* OS_printf("GPS_ReadAvailableData(): gps uart data len is 0\n"); */
+    }
+```
+
+### The NOS3 Driver
+
+The example described above uses the `hwlib.h` header which provides the device drivers for the Novatel OEM615. This header is included by any library making calls to the UART and can be stored at any location. In this case both the hardware library and device drivers are located in `fsw/apps/hwlib/fsw/public_inc/`.
+
+The functions used by the GPS app in this example are `uart_bytes_available` and `uart_read_port`, which are defined in `fsw/apps/hwlib/sim/src/libuart.c`. The `uart_bytes_available` function will be examined in more detail for this example. The code excerpt below shows the function.
+
+```c
+/* usart number bytes available */
+int32 uart_bytes_available(int32 handle)
+{
+    int bytes = 0;
+    NE_Uart *dev = nos_get_usart_device((int)handle);
+    if(dev)
+    {
+        OS_MutSemTake(nos_usart_mutex);
+        bytes = (int)NE_uart_available(dev);
+        OS_MutSemGive(nos_usart_mutex);
+    }
+    return bytes;
+}
+```
+
+The function above is used to return the number of bytes available from the USART buffer to the calling function. The `NE_uart_available` function used in this code is provided by NOS Engine. Details about the UART, I2C, and SPI NOS plugins can be found in the NOS Engine user’s manual.
+
+### Build System
+
+The build system must be able to properly select the correct driver source code based on the target being compiled. In this case, CMake is used by both cFS and NOS3 and can accomplish this swap easily. As described earlier the `targets.cmake` file in `fsw/nos3_defs` provides an example of how to include driver source code; an example Cmake build file may be found in `components/novatel_oem615/CmakeList.txt`.
```

Comment:
Change "may be found in `components/novatel_oem615/CmakeList.txt`"
to "may be found in `components/novatel_oem615/fsw/cfs/CMakeLists.txt"